src/com/linkedin/parseq/FusionTask.java

```diff
@@ -22,33 +22,61 @@
 
 
 /**
+ * FusionTask implements optimization in case when multiple synchronous
+ * transformations are applied in sequence. Instead of rescheduling every transformation
+ * as a separate task it is faster to apply transformations in a chain.
  *
- * @author jodzga
+ * This is internal class and it should not be extended or used outside ParSeq.
+ * It may change or be removed at any time in a backwards incompatible way.
  *
- * @param <S>
- * @param <T>
+ * @author Jaroslaw Odzga (jodzga@linkedin.com)
+ *
+ * @param <S> type of source
+ * @param <T> type of target
  */
-public class FusionTask<S, T> extends BaseTask<T> {
+class FusionTask<S, T> extends BaseTask<T> {
 
   private static final Logger LOGGER = LoggerFactory.getLogger(FusionTask.class);
 
   private static final Continuations CONTINUATIONS = new Continuations();
 
+  /* Encapsulates transformation from source to target, includes all predecessors' transformations, can't be null */
   private final Consumer3<FusionTraceContext, Promise<S>, Settable<T>> _propagator;
-  private final Task<S> _task;
+
+  /* Asynchronous task that will complete source promise, can be null in which case propagator must not depend on source */
+  private final Task<S> _asyncTask;
+
```

Comment:
It's a bit confusing why we need both a source task that will complete the source promise and the source promise itself. Can you add some text explaining why both of them are needed?